### DISCUSSION TITLE
how to use distance calculation in a loop.

### DISCUSSION THREAD

**Discussion by theo-allnutt-bioinformatics** [cogent3/cogent3#940] [Q&A] [answered]
Hi,

trying to calculate distances on many separate sequence alignments in a loop. If the there is not way to wait for the calculation to complete, how can it be used on more than one alignment?

e.g.

cds=fasta2dict(f2) #makes dictionary of all sequences
for i in f1:
	paralogs={}
	k=i.split("\t")
	k[-1]=k[-1].rstrip("\n")
	name1=k[0]
	paraloglist=k[2:] #list of paralogs to retrieve and align
	for j in paraloglist:
		paralogs[j]=cds[j]
	seqs = make_unaligned_seqs(data=paralogs, moltype="protein")
	aligned = aa_aligner(seqs)
	d = aligned.distance_matrix(calc="hamming", show_progress=False)
	print(d)
Traceback (most recent call last):
File "/g/data/nm31/scripts/paralog-dists.py", line 106, in
getgendists(f1,f2,g1)
File "/g/data/nm31/scripts/paralog-dists.py", line 80, in getgendists
d = aligned.distance_matrix(calc="hamming", show_progress=False)
AttributeError: 'NotCompleted' object has no attribute 'distance_matrix'

**Comment by GavinHuttley** [comment, marked as answer]
Using the latest version of the develop branch, the following will work (assuming you have a collection of fasta formatted DNA sequence files under a directory path `data/mixed_names`.

```python
from pathlib import Path
from cogent3.app import io, align, dist

mixed_names = Path("data/mixed_names")
dstore = io.get_data_store(mixed_names, suffix="fa*")

reader = io.load_unaligned(format="fasta")
aa_aligner = align.progressive_align("protein", unique_guides=True)
dist_calc = dist.fast_slow_dist(distance="hamming", moltype="protein")
writer = io.write_tabular(Path("results") / "mixed_names", create=True, if_exists="overwrite")

process = reader + aa_aligner + dist_calc + writer
r = process.apply_to(dstore)
```
And there will be a bunch of tsv files saved in the `results/mixed_names` directory.

Hope that helps!
